File: androidApp/core/compose/src/main/kotlin/com/velkonost/getbetter/core/compose/theme/Type.kt
package com.velkonost.getbetter.core.compose.theme

import androidx.compose.material3.Typography
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.TextUnitType
import androidx.compose.ui.unit.sp
import com.velkonost.getbetter.shared.resources.SharedR.fonts.sfpro

private val SFFontFamily = FontFamily(
    listOf(
        Font(resId = sfpro.rounded_regular.fontResourceId, weight = FontWeight.Normal),
        Font(resId = sfpro.rounded_semibold.fontResourceId, weight = FontWeight.SemiBold),
        Font(resId = sfpro.rounded_bold.fontResourceId, weight = FontWeight.Bold),
        Font(resId = sfpro.rounded_black.fontResourceId, weight = FontWeight.Black)
    )
)

/**
 * Values taken from Google Material 3 documentation
 * [Source](https://m3.material.io/styles/typography/type-scale-tokens)
 */
@Suppress("MagicNumber")
internal val AppTypography = Typography(
    // DISPLAY .....................................................................................
    displayLarge = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.Normal,
        fontSize = 57.sp,
        lineHeight = TextUnit(1.12F, TextUnitType.Em)
    ),
    displayMedium = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.Normal,
        fontSize = 45.sp,
        lineHeight = TextUnit(1.15F, TextUnitType.Em)
    ),
    displaySmall = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.Normal,
        fontSize = 36.sp,
        lineHeight = TextUnit(1.22F, TextUnitType.Em)
    ),

    // HEADLINE ....................................................................................
    headlineLarge = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.Normal,
        fontSize = 32.sp,
        lineHeight = TextUnit(1.25F, TextUnitType.Em)
    ),
    headlineMedium = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.Normal,
        fontSize = 28.sp,
        lineHeight = TextUnit(1.285F, TextUnitType.Em)
    ),
    headlineSmall = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.Normal,
        fontSize = 24.sp,
        lineHeight = TextUnit(1.33F, TextUnitType.Em)
    ),

    // TITLE .......................................................................................
    titleLarge = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.Black,
        fontSize = 22.sp,
        lineHeight = TextUnit(1.27F, TextUnitType.Em)
    ),
    titleMedium = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.Bold,
        fontSize = 16.sp,
        lineHeight = TextUnit(1.5F, TextUnitType.Em),
        letterSpacing = TextUnit(0.009f, TextUnitType.Em)
    ),
    titleSmall = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.SemiBold,
        fontSize = 14.sp,
        lineHeight = TextUnit(1.42F, TextUnitType.Em),
        letterSpacing = TextUnit(0.007f, TextUnitType.Em)
    ),

    // LABEL .......................................................................................
    labelLarge = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.SemiBold,
        fontSize = 16.sp,
        lineHeight = TextUnit(1.42F, TextUnitType.Em),
        letterSpacing = TextUnit(0.007f, TextUnitType.Em)
    ),
    labelMedium = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.SemiBold,
        fontSize = 14.sp,
        lineHeight = TextUnit(1.33F, TextUnitType.Em),
        letterSpacing = TextUnit(0.03f, TextUnitType.Em)
    ),
    labelSmall = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.SemiBold,
        fontSize = 11.sp,
        lineHeight = TextUnit(1.45F, TextUnitType.Em),
        letterSpacing = TextUnit(0.045f, TextUnitType.Em)
    ),

    // BODY ........................................................................................
    bodyLarge = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.SemiBold,
        fontSize = 16.sp,
        lineHeight = TextUnit(1.5F, TextUnitType.Em),
        letterSpacing = TextUnit(0.03f, TextUnitType.Em)
    ),
    bodyMedium = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.Normal,
        fontSize = 14.sp,
        lineHeight = TextUnit(1.42F, TextUnitType.Em),
        letterSpacing = TextUnit(0.015f, TextUnitType.Em)
    ),
    bodySmall = TextStyle(
        fontFamily = SFFontFamily,
        fontWeight = FontWeight.Normal,
        fontSize = 12.sp,
        lineHeight = TextUnit(1.33F, TextUnitType.Em),
        letterSpacing = TextUnit(0.033f, TextUnitType.Em)
    )
)

File: iosApp/iosApp/Core/Theme/Type.swift
//
//  Type.swift
//  iosApp
//
//  Created by velkonost on 10.09.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK

public extension Font {

    static let displayLarge = Font(SharedR.fontsSfpro.shared.rounded_regular.uiFont(withSize: 57))
    static let displayMedium = Font(SharedR.fontsSfpro.shared.rounded_regular.uiFont(withSize: 45))
    static let displaySmall = Font(SharedR.fontsSfpro.shared.rounded_regular.uiFont(withSize: 36))

    static let headlineLarge = Font(SharedR.fontsSfpro.shared.rounded_regular.uiFont(withSize: 32))
    static let headlineMedium = Font(SharedR.fontsSfpro.shared.rounded_regular.uiFont(withSize: 28))
    static let headlineSmall = Font(SharedR.fontsSfpro.shared.rounded_regular.uiFont(withSize: 24))

    static let titleLarge = Font(SharedR.fontsSfpro.shared.rounded_black.uiFont(withSize: 22))
    static let titleMedium = Font(SharedR.fontsSfpro.shared.rounded_bold.uiFont(withSize: 16))
    static let titleSmall = Font(SharedR.fontsSfpro.shared.rounded_bold.uiFont(withSize: 14))

    static let labelLarge = Font(SharedR.fontsSfpro.shared.rounded_semibold.uiFont(withSize: 16))
    static let labelMedium = Font(SharedR.fontsSfpro.shared.rounded_semibold.uiFont(withSize: 14))
    static let labelSmall = Font(SharedR.fontsSfpro.shared.rounded_semibold.uiFont(withSize: 11))

    static let bodyLarge = Font(SharedR.fontsSfpro.shared.rounded_semibold.uiFont(withSize: 16))
    static let bodyMedium = Font(SharedR.fontsSfpro.shared.rounded_regular.uiFont(withSize: 14))
    static let bodySmall = Font(SharedR.fontsSfpro.shared.rounded_regular.uiFont(withSize: 12))

}

public struct TextStyle {
    let font: FontResource
    let tracking: CGFloat
    let lineSpacing: CGFloat
    let size: CGFloat
    
    fileprivate init(_ font: FontResource, _ tracking: CGFloat, _ lineSpacing: CGFloat, _ size: CGFloat) {
        self.font = font
        self.tracking = tracking
        self.lineSpacing = lineSpacing
        self.size = size
    }
}

public extension TextStyle {
    
    static let displayLarge = TextStyle(SharedR.fontsSfpro.shared.rounded_bold, -2.5, 0, 57)
    static let displayMedium = TextStyle(SharedR.fontsSfpro.shared.rounded_regular, -2, 0, 45)
    static let displaySmall = TextStyle(SharedR.fontsSfpro.shared.rounded_regular, -1.5, 0, 36)
    
    static let headlineLarge = TextStyle(SharedR.fontsSfpro.shared.rounded_regular, -1, 0, 32)
    static let headlineMedium = TextStyle(SharedR.fontsSfpro.shared.rounded_regular, -0.5, 0, 28)
    static let headlineSmall = TextStyle(SharedR.fontsSfpro.shared.rounded_regular, -0.75, 0, 24)
    
    static let titleLarge = TextStyle(SharedR.fontsSfpro.shared.rounded_black, -0.1, 0, 22)
    static let titleMedium = TextStyle(SharedR.fontsSfpro.shared.rounded_bold, 0, 0, 16)
    static let titleSmall = TextStyle(SharedR.fontsSfpro.shared.rounded_semibold, 0.1, 0, 14)
    
    static let labelLarge = TextStyle(SharedR.fontsSfpro.shared.rounded_semibold, 0.1, 0, 16)
    static let labelMedium = TextStyle(SharedR.fontsSfpro.shared.rounded_semibold, 0.5, 0, 14)
    static let labelSmall = TextStyle(SharedR.fontsSfpro.shared.rounded_semibold, 0.5, 0, 11)
    
    static let bodyLarge = TextStyle(SharedR.fontsSfpro.shared.rounded_semibold, 0.5, 4, 16)
    static let bodyMedium = TextStyle(SharedR.fontsSfpro.shared.rounded_regular, 0.25, 3, 14)
    static let bodySmall = TextStyle(SharedR.fontsSfpro.shared.rounded_regular, 0.4, 2, 12)
}
